### PR TITLE
Tools: ardupilotwaf: Lazily evaluate default for `os.environ.get(key,default)`

### PR DESCRIPTION
**Pull request by balattar**
## Description
When parsing for the git hash of the current branch (which is required for the waf build), the call to os.environ first looks for a `GIT_VERSION` key in the environment. If it fails, it returns the output `git rev-parse HEAD` for the git version. 
 
In some build environments, access to the `.git` folder is limited when copying over the repository, which results in an error and failed build when trying to run `git rev-parse`. Setting the `GIT_VERSION` and `CHIBIOS_GIT_VERSION` environment variables allows the user to circumvent this, but since the default is eagerly evaluated, the build will still throw an error and terminate.

This PR replaces the calls to `os.environ.get(value, default)` with lazy functions that only evaluate the default (with git) if the environment variable is not set.

### CODE REVIEW

File: Tools/ardupilotwaf/ardupilotwaf.py [hunk 1]
```diff
@@ -373,7 +373,7 @@ def ap_find_tests(bld, use=[]):
 
 @conf
 def ap_version_append_str(ctx, k, v):
-    ctx.env['AP_VERSION_ITEMS'] += [(k, '"{}"'.format(os.environ.get(k, v)))]
+    ctx.env['AP_VERSION_ITEMS'] += [(k, '"{}"'.format(os.environ.get(k) if k in os.environ else v)))]
```
Comment (peterbarker) [inline Tools/ardupilotwaf/ardupilotwaf.py:376]: Isn't this a noop?  There's no evaluation going on here as you've described in the PR.

File: Tools/ardupilotwaf/chibios.py [hunk 2]
```diff
@@ -205,7 +205,7 @@ def run(self):
         desc_len = 16
         crc1 = to_unsigned(crc32(bytearray(img[:offset])))
         crc2 = to_unsigned(crc32(bytearray(img[offset+desc_len:])))
-        githash = to_unsigned(int('0x' + os.environ.get('GIT_VERSION', self.generator.bld.git_head_hash(short=True)),16))
+        githash = to_unsigned(int('0x' + os.environ.get('GIT_VERSION') if 'GIT_VERSION' in os.environ else self.generator.bld.git_head_hash(short=True),16))
```
Comment (peterbarker) [inline Tools/ardupilotwaf/chibios.py:208]: We don't tend to use this if/else format; please split this across multiple statements.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: Tools/ardupilotwaf/ardupilotwaf.py (hunk 1 above)
Comment (bma5287) [reply Tools/ardupilotwaf/ardupilotwaf.py]: Yup my bad. Moving this up to the `ap_version_append_str` call.

File: Tools/ardupilotwaf/boards.py [hunk 3]
```diff
@@ -473,8 +473,18 @@ def pre_build(self, bld):
             self.embed_ROMFS_files(bld)
 
     def build(self, bld):
-        bld.ap_version_append_str('GIT_VERSION', bld.git_head_hash(short=True))
-        bld.ap_version_append_int('GIT_VERSION_INT', int("0x" + bld.git_head_hash(short=True), base=16))
+        version_type = 'GIT_VERSION'
+        if version_type in os.environ:
+            bld.ap_version_append_str(version_type, version_type)
```
Comment (alkasm) [inline Tools/ardupilotwaf/boards.py:478]: bld.ap_version_append_str(version_type, os.environ[version_type])

File: Tools/ardupilotwaf/boards.py [hunk 4]
```diff
@@ -473,8 +473,18 @@ def pre_build(self, bld):
...
+        else:
+            bld.ap_version_append_str(version_type, bld.git_head_hash(short=True))
+            
+        version_type_int = 'GIT_VERSION_INT'
+        if version_type_int in os.environ:
+            bld.ap_version_append_str(version_type_int, version_type_int)
```
Comment (alkasm) [inline Tools/ardupilotwaf/boards.py:484]: same

File: Tools/ardupilotwaf/boards.py [hunk 5]
```diff
@@ -1074,7 +1084,11 @@ def configure_env(self, cfg, env):
 
     def build(self, bld):
         super(chibios, self).build(bld)
-        bld.ap_version_append_str('CHIBIOS_GIT_VERSION', bld.git_submodule_head_hash('ChibiOS', short=True))
+        version_type = 'CHIBIOS_GIT_VERSION'
+        if version_type in os.environ:
+            bld.ap_version_append_str(version_type, version_type)
```
Comment (alkasm) [inline Tools/ardupilotwaf/boards.py:1089]: same

File: Tools/ardupilotwaf/boards.py (hunk 3 above)
Comment (alkasm) [inline Tools/ardupilotwaf/boards.py:478]:
```suggestion
            bld.ap_version_append_str(version_type, os.environ[version_type])
```

File: Tools/ardupilotwaf/boards.py (hunk 4 above)
Comment (alkasm) [inline Tools/ardupilotwaf/boards.py:484]:
```suggestion
            bld.ap_version_append_str(version_type_int, os.environ[version_type_int])
```

File: Tools/ardupilotwaf/boards.py (hunk 5 above)
Comment (alkasm) [inline Tools/ardupilotwaf/boards.py:1089]:
```suggestion
            bld.ap_version_append_str(version_type, os.environ[version_type])
```

File: Tools/ardupilotwaf/chibios.py [hunk 6]
```diff
@@ -283,7 +283,11 @@ def run(self):
         img2 = bytearray(img[offset+desc_len:])
         crc1 = to_unsigned(crc32(img1))
         crc2 = to_unsigned(crc32(img2))
-        githash = to_unsigned(int('0x' + os.environ.get('GIT_VERSION', self.generator.bld.git_head_hash(short=True)),16))
+        if 'GIT_VERSION' in os.environ:
+            hex_hash = os.environ.get('GIT_VERSION')
+        else:
+            hex_hash = self.generator.bld.git_head_hash(short=True)
+        githash = to_unsigned(int('0x' + hex_hash, 16))
```
Comment (alkasm) [inline Tools/ardupilotwaf/chibios.py:290]: unrelated to your change but the `"0x"` is superfluous: `int(s, 16) == int("0x" + s, 16)`

File: Tools/ardupilotwaf/boards.py [hunk 7]
```diff
@@ -473,8 +473,18 @@ def pre_build(self, bld):
...
+        if version_type in os.environ:
+            bld.ap_version_append_str(version_type, os.environ[version_type])
+        else:
+            bld.ap_version_append_str(version_type, bld.git_head_hash(short=True))
+            
+        version_type_int = 'GIT_VERSION_INT'
+        if version_type_int in os.environ:
+            bld.ap_version_append_str(version_type_int, os.environ[version_type_int])
+        else:
+            bld.ap_version_append_str(version_type_int, int("0x" + bld.git_head_hash(short=True), base=16))
```
Comment (bma5287) [inline Tools/ardupilotwaf/boards.py:486]: Should use `ap_version_append_int`

```suggestion
        if version_type_int in os.environ:
            bld.ap_version_append_int(version_type_int, os.environ[version_type_int])
        else:
            bld.ap_version_append_int(version_type_int, int("0x" + bld.git_head_hash(short=True), base=16))
```

**Comment by bma5287** [comment]
This is missing one more eval at 
https://github.com/ArduPilot/ardupilot/blob/b0f0733ec342657d3b9d85d991f8ca3dfb228c2a/Tools/ardupilotwaf/chibios.py#L331

**Pull request closed without merging**